interfaces/apparmor/apparmor.go

```diff
@@ -83,22 +85,49 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
 	return nil
 }
 
-// UnloadProfiles removes the named profiles from the running kernel.
-//
-// The operation is done with: apparmor_parser --remove $names...
-// The binary cache file is removed from /var/cache/apparmor
+// UnloadProfiles is meant to remove the named profiles from the running
+// kernel and then remove any cache files. Importantly, we can only unload
+// profiles when we are sure there are no lingering processes from the snap
+// (ie, forcibly stop all running processes from the snap). Otherwise, any
+// running processes will become unconfined. Since we don't have this guarantee
+// yet, leave the profiles loaded in the kernel but remove the cache files from
+// the system so the policy is gone on the next reboot. LP: #1818241
 func unloadProfiles(names []string, cacheDir string) error {
 	if len(names) == 0 {
 		return nil
 	}
 
-	args := []string{"--remove"}
-	args = append(args, names...)
-	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("cannot unload apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+	/* TODO: uncomment when no lingering snap processes is guaranteed
+	// By the time this function is called, all the profiles (names) have
+	// been removed from dirs.SnapAppArmorDir, so to unload the profiles
+	// from the running kernel we must instead use sysfs and write the
+	// profile names one at a time to
+	// /sys/kernel/security/apparmor/.remove (with no trailing \n).
+	apparmorSysFsRemove := "/sys/kernel/security/apparmor/.remove"
+	if !osutil.IsWritable(appArmorSysFsRemove) {
+	        return fmt.Errorf("cannot unload apparmor profile: %s does not exist\n", appArmorSysFsRemove)
 	}
-	if err := osutil.UnlinkMany(cacheDir, names); err != nil && !os.IsNotExist(err) {
+	for _, n := range names {
+	        // ignore errors since it is ok if the profile isn't removed
+	        // from the kernel
+	        ioutil.WriteFile(appArmorSysFsRemove, []byte(n), 0666)
+	}
+	*/
+
+	// AppArmor 2.13 and higher has a cache forest while 2.12 and lower has
```

Comment:
I was thinking about this fragment and given that I don't fully understand the 2.13 format yet I wanted to change how we handle this code. I opened a mostly-documentation RFC pull request at https://github.com/snapcore/snapd/pull/6559

Reply:
> I was thinking about this fragment and given that I don't fully understand the 2.13 format yet I wanted to change how we handle this code.

I don't understand: you say you don't understand the format but yet you want to change it?

Reply:
I had a similar reaction and felt like this should be refactored. But given how small and isolated this is I think its ok for this PR and we can improve/iterate in a followup.

Reply:
Refactoring in a followup is of course fine, but yeah, I didn't do that precisely because this only needed to be done in this one spot. I also suspect the cache forest to be in place for many years (ie, I don't foresee it changing, so less of a reason to abstract this out).